src/storage/index/hnsw_index.cpp

```diff
@@ -339,37 +338,272 @@ common::offset_t OnDiskHNSWIndex::searchNNInUpperLayer(transaction::Transaction*
+    }
+    while (!candidates.empty()) {
+        auto [candidate, candidateDist] = candidates.top();
+        // Break here if adding closestNode to result will exceed efs or not improve the results.
+        if (results.size() >= ef && candidateDist > results.top().distance) {
+            break;
+        }
+        candidates.pop();
+        auto neighbors = searchFilteredNeighbors(transaction, candidate, queryVector, searchType,
+            searchState, *scanState, *secondHopScanState, ef, candidates, results);
+        for (auto nbr : neighbors) {
+            const auto nbrVector = embeddings->getEmbedding(transaction,
+                *searchState.embeddingScanState.scanState, nbr);
+            dist = HNSWIndexUtils::computeDistance(config.distFunc, queryVector, nbrVector,
+                embeddings->getDimension());
+            if (results.size() < ef || dist < results.top().distance) {
+                if (results.size() == ef) {
+                    results.pop();
+                }
+                results.push({nbr, dist});
+                candidates.push({nbr, dist});
+            }
+        }
+    }
+
+    return popTopK(results, searchState.k);
+}
+
+common::offset_vec_t OnDiskHNSWIndex::searchFilteredNeighbors(transaction::Transaction* transaction,
+    common::offset_t nodeOffset, const float* queryVector, FilteredSearchType searchType,
+    HNSWSearchState& searchState, graph::NbrScanState& nbrScanState,
+    graph::NbrScanState& secondHopNbrScanState, uint64_t ef, min_node_priority_queue_t& candidates,
+    max_node_priority_queue_t& results) const {
+    auto neighborItr = lowerGraph->scanFwd(common::nodeID_t{nodeOffset, nodeTableID}, nbrScanState);
+    common::offset_vec_t neighbors;
+    switch (searchType) {
+    case FilteredSearchType::BLIND_TWO_HOP: {
+        neighbors = blindTwoHopFilteredSearch(searchState.semiMask, ef, neighborItr,
+            searchState.visited, secondHopNbrScanState);
+    } break;
+    case FilteredSearchType::DIRECTED_TWO_HOP: {
+        neighbors = directedTwoHopFilteredSearch(transaction, searchState.semiMask, queryVector, ef,
+            neighborItr, searchState.visited, *searchState.embeddingScanState.scanState,
+            secondHopNbrScanState, candidates, results);
+    } break;
+    case FilteredSearchType::ONE_HOP: {
+        neighbors =
+            oneHopFilteredSearch(searchState.semiMask, ef, neighborItr, searchState.visited);
+    } break;
+    default: {
+        KU_UNREACHABLE;
+    }
+    }
+    return neighbors;
+}
+
+common::offset_vec_t OnDiskHNSWIndex::oneHopFilteredSearch(common::SemiMask* semiMask, uint64_t ef,
+    graph::Graph::EdgeIterator& nbrItr, VisitedState& visited) {
+    common::offset_vec_t result;
+    result.reserve(ef);
+    for (const auto& neighborChunk : nbrItr) {
+        neighborChunk.forEach([&](auto neighbor, auto) {
+            if (!visited.contains(neighbor.offset)) {
```

Comment:
We could simplify this to if (!visited.contains(neighbor.offset) && semiMask->isMasked(neighbor.offset)) { }.

Reply:
Done.